### DISCUSSION TITLE
Gremlin, automatic use of indexes on where() steps

### DISCUSSION THREAD

**Discussion by marco-brandizi** [ArcadeData/arcadedb#1168] [General]
Hi all,

I thought of asking ArcadeDB people on [this issue](https://stackoverflow.com/q/76527659/529286), after having realised that maybe, the core of the problem is about using indexes with Gremlin. 

To summarise it, I have edges that are 'ternary relations', ie, in addition to the usual outgoing/incoming vertexes, the relation has an 'evidence' attribute, which contains the ID of a third 'EvidenceType' vertex (which could describe something like: manually curated, imported from &lt;ref&gt;, text mining).

The intuitive query for this (see [my own answer](https://stackoverflow.com/a/76527797/529286)) is traversing the edges of interest, adding another traversal over `V()` and using `where()` to match the edge's attribute to the vertex ID in the second traversal. This is pretty similar to multiple JOINs in SQL or multiple MATCHes in Cypher. However, in ArcadeDB, it seems to be resolving it with a full scan of the evidence type, for each matched edge in the second part. Apparently, it doesn't use the indexes (neither the one on vertext ID, nor the one on the edge property), nor it caches already-found vertexes (on 100 edges, there are just 4-5 linked evidences).

I understand that's usually the way Gremlin is implemented, but I'd like to investigate more: is there a way to tell the Gremlin engine to use the indexes? Is there a way to write that query in pure Gremlin, without using the map-based approach that I discovered later (again see the last part of my answer)? That approach isn't good when the in-memory intermediate result that you need to accumulate is too big.

Am I missing something in Gremlin (I'm quite new to it)?

Note that "model it differently" isn't possible in this case, and also I'm investigating the potential of Gremlin with various data models, including this one, which occurs quite often in my application domain.

**Comment by lvca** [comment]
Hi @marco-brandizi, I'd like to check what's going on under the hood and why the index is not used. Is there a way where I can have a minimal database where to run your query and see what's going on? Even the database with 1 record for type would be enough.

**Reply by marco-brandizi** [reply]
Thanks @lvca 

Please, could you try [this backup](https://app.box.com/s/zbfxv6ojj5z2hk7nnksjhrinkhfljmxt)? I've made it with the ArcadeDB Backup command in the Studio, I'm using v23.4.1. It should include all the indexes I've created with SQL.

You'll notice that you can't see vertex type details on the studio, due to [this bug](https://github.com/ArcadeData/arcadedb/issues/1169) I've just filed. They work fine nonetheless.

Many thanks!

**Reply by lvca** [reply]
The issue is with ArcadeDB query optimizer and the `contains` operator. If remove contains and just use the label names, it would use the index and should return in <10ms:

```gremlin
evLabels = [:]
g.V().hasLabel ( 'Concept:Protcmplx:Resource' ).as ( 'cpx' )
  // Trying to put the limit early-on
  .inE().hasLabel ( 'is_part_of' ).limit ( 100 ).as ( 'pr' )
  .outV ().hasLabel ( 'Concept:Protein:Resource' ).as ( 'p' )
.select ( 'p', 'cpx', 'pr' )
  .by ( 'prefName' )
  .by ( 'prefName' )
  .by ( map{
    pr = it.get()
    evIri = pr.values ( 'evidence' ).next ();
    lbl = evLabels [ evIri ];
    if ( lbl != null ) return lbl;
    lbl = g.V().hasLabel ( 'EvidenceType:Resource' )
             .has ( 'iri', evIri )
             .values ( 'label' ).next ();
    evLabels [ evIri ] = lbl == null ? "" : lbl;
    return lbl;
  })
```

**Reply by lvca** [reply]
This takes only 7ms on my Mac. The `contains` would skip any optimization with fetching directly from the label (type).

**Reply by lvca** [reply]
Also I noticed, by just looking at the names, that maybe you could use inheritance here. For example, the type `Concept:Protein:Resource` could extend the types `Concept` and `Resource`. In this way you could execute polymorphic queries, like `hasLabel( 'Concept' )` to query from all the concept instances and its subtypes.

**Reply by marco-brandizi** [reply]
Thanks again, as I wrote on SO, the variant with label exact matching doesn't make things much faster (though it gains some ms), for the major issue is on linking edge.evidence to EvidenceType.iri.

Using this little improvement on the query that collects intermediate on a map helps, but just a little (7-10ms for 10 results is in line with my 100ms for 100 results, and it confirms this version scales linearly), using such improvement with the variant that joins with where() still yields about 2s for 10 results and the profile() step shows the same full scan problem.

As for the inheritance, interesting, but I'm coming from graphML data that aren't ArcadeDB-specific [[REF](https://github.com/Rothamsted/rdf2pg)].

**Reply by marco-brandizi** [reply]
Sorry, I've reviewed it again and indeed, the map-based query with label exact matching is 10x faster. However, the version with where() has more or less the same speed.

**Reply by lvca** [reply]
For reference, is this the query?

```gremlin
g.V().hasLabel ( containing ( 'Protcmplx' ) ).as ( 'cpx' )
  .inE().hasLabel ( 'is_part_of' ).limit ( 10 ).as ( 'pr' )
  .outV ().hasLabel ( containing ( 'Protein' ) ).as ( 'p' )  
.V().hasLabel ( containing ( 'EvidenceType' ) ).as ( 'ev' )
    .where ( 'ev', eq ( 'pr' ) ).by ( 'iri' ).by ( 'evidence' ) 
.select ( 'p', 'cpx', 'ev' )
.by ( 'prefName' )
.by ( 'prefName' )
.by ( 'label' )
```

Is there a particular reason why you do `.inE().hasLabel ( 'is_part_of' )` instead of `.inE('is_part_of')`? Also, have you tried using `has('ev', 'pr' )` instead of `where ( 'ev', eq ( 'pr' ) )`?

**Reply by lvca** [reply]
I'm not super skilled with Gremlin. I am trying to understand what you want to extract from the database to do the same in SQL.

I started with this:

```sql
select $pr, $p
from `Concept:Protcmplx:Resource`
let pr = inE('is_part_of'), 
     p = $pr.outV()[@type = 'Concept:Protein:Resource']
```

But then you're continuing with `.V()` to get evidence type by using where with the by and here I'm lost :-)

**Reply by marco-brandizi** [reply]
It's surprising, cause to me this is a quite common case and instead, it seems hard to describe here or on SO, in addition to the fact I can't find a similar case in Gremlin documentation. 

Anyway, this is how the data look like I've tried to describe them in SO:

```
v1: Protein{prefName: 'QP1'} 
  -- r1: is_part_of{evidence: 'ns:testdb'} 
  --> v2: Protcmplx{prefName: 'P12 Complex'}
ev: EvidenceType{ iri = "ns:testdb", label = "Test Database" }
```

that is, each is_part_of edge links a Protein vertex to a Protcmplx vertex (ie, it's saying the protein is part of some aggregate of proteins), then, conceptually a third link goes from the same is_part_of edge to an EvidenceType vertex. The way the latter is modelled is that the edge's property 'evidence' is the same as the EvidenceType's property 'iri' (primary key for vertexes).

So, what I'm trying to do is listing tuples of Protein(prefName)/complex(prefName)/evidence(label).

So far, the only way I found to do so in Gremlin is:

* either collect unique 'evidence' values from the edges and then use this smaller list to query EvidenceType(s) in a lambda
 
* or, use two traversals, one for the part-of part and the other for EvidenceType and then join the two by matching the corresponding is_part_of.evidence/EvidenceType.iri.

The former is very fast, but it creates a map in-memory, so not very good if there are many expected evidences as results. The second is still slow, presumably cause it does a full scan of evidences for each is_part_of that it meets.

In Cypher, it works like this: 

```
MATCH (cpx:Protcmplx) <- [ pr:is_part_of ] - (p:Protein), (ev:EvidenceType)
WHERE pr.evidence = ev.iri
RETURN pr, ev
LIMIT 10
```
which can be tried [here](http://knetminer-wheat.cyverseuk.org:7474/browser/) (leave credentials blank) and it is fast enough, despite it's against a much bigger dataset.

Coming to your proposed changes (thank you so much, very instructive!):

* `inE('is_part_of')` makes things a bit faster, but doesn't change much on the full-scan/cartesian that seem to be happening in `where()`

* `has()` can't work as you say, cause I need to match the attributes, so I've tried this instead:

```
g.V().hasLabel ( 'Concept:Protcmplx:Resource' ).as ( 'cpx' )
  .inE('is_part_of').limit ( 10 ).as ( 'pr' )
  .outV ().hasLabel ( 'Concept:Protein:Resource' ).as ( 'p' )  
.V().hasLabel ( 'EvidenceType:Resource' ).as ( 'ev' )
  .has ( 'iri', select ( 'pr' ).by ( 'evidence' ) )
.select ( 'p', 'cpx', 'ev' )
.by ( 'prefName' )
.by ( 'prefName' )
.by ( 'label' )
```

But it's still around 1.2s for 10 results.

Thanks again!

**Reply by lvca** [reply]
Sorry, I think I didn't read the whole SO thread then. Got it, you are building a ternary relationship by using the edges. With ArcadeDB you could put a link to the EvidenceType, so you don't have to do the join, but just traverse it from the edge. By default, an edge has `@in` and `@out`, but you could have multiple links to other vertices, edges, or documents.

Without such change, the same cypher query in ArcadeDB would be using a sub query:

```sql
select $resource, $relationship, $evidence
from `Concept:Protcmplx:Resource`
let relationship = inE('is_part_of'), 
    resource = $relationship.outV()[@type = 'Concept:Protein:Resource'],
    evidence = ( select from `EvidenceType:Resource` where iri in $parent.relationship.evidence )
limit 10
```

This takes about 27ms on my Mac (please update to the latest `main` branch to have the latest fix on studio).

But if you're able to convert the `evidence` property in the edge to the actual link to the Evidence (I guess during the import phase), then the query would be much faster and like this:

```sql
select $resource, $relationship, $evidence
from `Concept:Protcmplx:Resource`
let relationship = inE('is_part_of'), 
    resource = $relationship.outV()[@type = 'Concept:Protein:Resource'],
    evidence = ( select from $parent.relationship.evidence )
limit 10
```

**Reply by marco-brandizi** [reply]
That's very good to know, though for the moment I'm trying to assess how Gremlin is easy-to-use and performant, with data models that are common in life sciences. Essentially, I and other collaborators are working on an extension of [this work](https://figshare.com/articles/dataset/Getting_the_best_of_Linked_Data_and_Property_Graphs_rdf2neo_and_the_KnetMiner_Use_Case/7314323). By the way, if you are interested, I'd be happy to collaborate on it with ArcadeDB people!

**Reply by lvca** [reply]
Absolutely! Is the dataset the same as you shared? Do you have Neo4j's cypher queries to convert and give them a spin on ArcadeDB?

**Reply by lvca** [reply]
Also, by looking at this diagram, you use edge types (like `part_of` and `consumed_by`) for connecting different vertex types. 

![image](https://github.com/ArcadeData/arcadedb/assets/312606/ce9d6a94-c676-424d-9e0e-bdb5d7f82af4)

This can be highly optimized in ArcadeDB by defining specific edge types:

```sql
create edge type part_of
create edge type transport_part_of extends part_of
create edge type reaction_part_of extends part_of
```

Then use `transport_part_of` for edges from the Transport vertex and `reaction_part_of` from Reaction. With polymorphism, you can still use the generic "part_of", but now you can traverse directly by the sub edge type instead of checking the target vertex. Ea

Like this part:

```sql
inE('is_part_of').outV()[@type = 'Concept:Protein:Resource']
```

Could be just written as:

```sql
inE('protein_is_part_of').outV()
```

If you use the `protein_is_part_of` only to map Protcmplx to Protein avoiding to fetch the vertex and checking for the type.
